src/core/actions/cityactions/Spawn.java

```diff
@@ -23,19 +28,48 @@ public Spawn(City c)
 
     @Override
     public LinkedList<Action> computeActionVariants(GameState gs) {
-        //TODO: compute spawn actions
-        return null;
+        LinkedList<Action> actions = new LinkedList<>();
+
+        Tribe tribe = gs.getTribe(city.getTribeId());
+        TechnologyTree t = tribe.getTechTree();
+        int star = tribe.getStars();
+        if (gs.getBoard().getUnitAt(city.getX(), city.getY()) == null){
+            for(Types.UNIT unit: Types.UNIT.values()){
+                // Only can create land units
+                if (unit.getCost() <= star && unit.getKey() <= 7){
+                    if (unit.getRequirement() == null){
+                        Spawn newAction = new Spawn(city);
+                        newAction.setUnitType(unit);
+                        actions.add(newAction);
+                    }else if(t.isResearched(unit.getRequirement())){
+                        Spawn newAction = new Spawn(city);
+                        newAction.setUnitType(unit);
+                        actions.add(newAction);
+                    }
+                }
+            }
+        }
+        return actions;
     }
 
     @Override
     public boolean isFeasible(GameState gs) {
-        //TODO: Is feasible spawn
-        return false;
+        boolean costRequirement =  gs.getTribe(city.getTribeId()).getStars() >= unit_type.getCost();
+        boolean techniqueRequirement = true;
+        if (unit_type.getRequirement() != null){
+            techniqueRequirement = gs.getTribe(city.getTribeId()).getTechTree().isResearched(unit_type.getRequirement());
+        }
+        return costRequirement && techniqueRequirement;
     }
 
     @Override
     public boolean execute(GameState gs) {
-        //TODO: execute a spawn action.
+        if (isFeasible(gs)){
+            Unit newUnit = unit_type.createUnit(new Vector2d(city.getX(), city.getY()), 0, false, city.getActorID(), city.getTribeId(), unit_type);
+            gs.getTribe(city.getTribeId()).subtractStars(unit_type.getCost());
+            gs.getBoard().setUnitIdAt(city.getX(), city.getY(), newUnit);
```

Comment:
Line 70: rather than setUnitIdAt(x,y,u), we should use 2 calls:
                        board.addUnitToBoard(unit);
                        board.addUnitToCity(unit, city);
This assigns an Agent ID to the unit, sets the unit in the board and adds the unit_id to the city.

I'm aware this is not very sustainable, we need to think of a way to add a unit with a single method call.  @rovenr12